murenn/dtcwt/nn.py

```diff
@@ -69,6 +72,84 @@ def forward(self, x):
             UWx_j = UWx_j.view(B, self.C, self.Q, N)
             output.append(UWx_j)
         return torch.stack(output, dim=3)
+    
+    @property
+    def to_conv1d(self):
+        """
+        Compute the single-resolution equivalent impulse response of the MuReNN layer.
+        This would be helpful for visualization in Fourier domain, for receptive fields,
+        and for comparing computational costs.
+           DTCWT        conv1d        IDTCWT
+        δ -------> ψ_j --------> w_jq -------> y_jq
+        -------
+        Return:
+            conv1d (torch.nn.Conv1d): A Pytorch Conv1d instance with weights initialized to y_jq.
+        Examples
+        --------
+        >>> import matplotlib.pyplot as plt
+        >>> tfm = murenn.MuReNNDirect(J=8, Q=5, T=32, in_channels=1)
+        >>> conv1d = tfm.to_conv1d
+        >>> x = torch.zeros(1,1,2**10)
+        >>> x[0,0,N//2]=1
+        >>> x = x*(1+0j)
+        >>> w = conv1d(x).reshape(J,Q,-1).detach()
+        >>> colors = [
+        >>>     'tab:blue', 'tab:orange', 'tab:green', 'tab:red', 'tab:purple',
+        >>>     'tab:brown', 'tab:pink', 'tab:gray', 'tab:olive', 'tab:cyan']
+        >>> for j in range(J):
+        >>>     for q in range(Q):
+        >>>         plt.semilogx(torch.abs(torch.fft.fft(w[j,q,:])), color=colors[j])
+        >>> plt.xlim(0, N//2)
+        """
+        # T the filter length
+        T = self.conv1d[0].kernel_size[0]
+        # J the number of levels of decompostion
+        J = self.dtcwt.J
+        # Generate the impulse signal
+        N = 2**J * T
+        x = torch.zeros(1, self.C, N)
+        x[:, :, N//2] = 1
+        # Get the padding mode
+        padding_mode = self.dtcwt.padding_mode
+        if padding_mode == "constant":
+            padding_mode = "zeros"
+
+        inv = murenn.IDTCWT(
+            J=J,
+            padding_mode=padding_mode,
+            normalize=False,
+        )
+        # Get DTCWT impulse reponses
+        phi, psis = self.dtcwt(x)
+        # Set phi to a zero valued tensor
+        zeros_phi = phi.new_zeros(size=(1, self.C*self.Q, phi.shape[-1]))
+        # Create an empty list for {w_jq}
+        ws = []
+        for j in range(J):
+            # Wpsi_jr = Re[psi_j] * w_jq
+            Wpsi_jr = self.conv1d[j](psis[j].real)
+            # W_ji = Im[psi_j] * w_jq
+            Wpsi_ji = self.conv1d[j](psis[j].imag)
+            # Set the coefficients besides this scale to zero
+            Wpsis_jr = [Wpsi_jr * (1 + 0j) if k == j else psis[k].new_zeros(size=psis[k].shape).repeat(1, self.Q, 1) for k in range(J)]
+            Wpsis_ji = [Wpsi_ji * (0 + 1j) if k == j else psis[k].new_zeros(size=psis[k].shape).repeat(1, self.Q, 1) for k in range(J)]
```

Comment:
why convert to complex? and why multiply by `1j`?

Reply:
I converted them to complex because you wanted the to_conv1d method to return a complex-valued PyTorch tensor in #45.   

Besides, in dtcwt, we have:  (I've uploaded a test to show this. )

$$
\begin{align}
idtcwt(x) &= idtcwt.tree_a(x) + idtcwt.tree_b(x)\\
&= idtcwt(x.real) + idtcwt(x.imag) 
\end{align}
$$


I think in the returned `Conv1d` instance, `conv1d.weight.real+conv1d.weight.imag` is equivalent to the case if we don't convert `Wpsi_j` to complex

there is a bug: maybe i should use $x=(1-1j)*\delta$ instead of $x=(1+0j) * \delta$ to convolve with `conv1d` to visualize the filter

I can change this into a real-valued version if this is too complicated

Reply:
Hello, i don't think $x = (1 - 1j) \delta$ is necessary. $x = \delta$ should suffice.

My question was not a mathematical question but more like a CS question. Usually when you do
`z = torch.complex(re, im)`
`re` and `im` are both real-valued. Here this is not the case, since you have multiplied `im` by `1j`.

Reply:
Thanks for your advice, I've made the change.  
I was not clear in my previous answer. In the algorithm, only the real part of a complex tensor would be passed through Tree A, and only the imaginary part would be passed through Tree B. That's why I was converging `z` into a complex-valued tensor.